# Demo of AI agent output

In [1]:
import openai
import json
import os

In [2]:
# Function to load text from a JSON file
def load_text_from_json(file_path):
    with open(file_path, 'r') as f:
        data = json.load(f)
    return data

# Function to call GPT-4 with the provided prompt and appended text
def call_gpt_4(client, messages):
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=messages,
        max_tokens=1500  # Adjust max_tokens based on your needs
    )
    
    return response.choices[0].message.content.strip()

In [3]:
json_file_path = "award_sample.json"

# Load the API key 
with open(os.path.expanduser("~/.openai")) as f:
    line = f.readline().strip()
    if line.startswith("api_key="):
        api_key = line.split("=")[1]
if not api_key:
    raise ValueError("API key not found. Please set the OPENAI_API_KEY environment variable.")

## Persona 1 – Dr. Paul the Principal Investigator
- Sample prompts or information Dr. Paul, a Principal Investigator and future proposer may want to know
- “I want to submit a proposal to NSF to understand and process animal sounds and language. Is my idea or topic already funded at NSF? Which program and program officer can I talk to about my idea?”

In [4]:
# Create OpenAI client
client = openai.OpenAI(api_key=api_key)

# Load the text from the JSON file
appended_text = load_text_from_json(json_file_path)
prompt = "I want to submit a proposal to NSF to understand and process animal sounds and language. Is my idea or topic already funded at NSF? Which program and program officer can I talk to about my idea?"

messages = [
    {
        "role": "system",
        "content": (
            "You are a helpful assistant who always follows requests exactly. "
            "You review the request and reflect upon your response before returning it to the user. "
            "The source data is from the NSF Award Search website. "
            "If the award data contains references, you should research those references and include the relevant information in your response."
            "If you use any references, please include them as citations in the response."
        )
    },
    {
        "role": "user",
        "content": f"{prompt}\nJSON data:{appended_text}"
    }
]

# Call GPT-4 and get the first response
response = call_gpt_4(client, messages)

print("AI Agent Response:")
print(response)

AI Agent Response:
Your proposed idea of understanding and processing animal sounds and language does indeed align with multiple topics supported by the NSF, specifically within the scope of the Animal Behavior, Population & Community Ecology, Ocean Tech & Interdiscip Coordin, and related programs. Based on the JSON data provided, there are several funded projects that fall under similar thematic concepts. Here’s an overview that might be beneficial to your proposal:

1. **Title**: REU Site: Animal Language Processing and Understanding
   - **ID**: 2349713
   - **Awardee**: University of Texas at Arlington
   - **Summary**: This project investigates whether animals have languages that are syntactically and semantically compositional like humans, using advanced machine learning.
   - **Program Officer**: Sharmistha Bagchi-Sen
   - **NSF Programs**: Research Experiences for Undergraduates (REU)

2. **Title**: STTR Phase I: Using Audio Analytics and Sensing to Enhance Broiler Chicken Welf

AI Agent Response:
Your proposed idea of understanding and processing animal sounds and language does indeed align with multiple topics supported by the NSF, specifically within the scope of the Animal Behavior, Population & Community Ecology, Ocean Tech & Interdiscip Coordin, and related programs. Based on the JSON data provided, there are several funded projects that fall under similar thematic concepts. Here’s an overview that might be beneficial to your proposal:

1. **Title**: REU Site: Animal Language Processing and Understanding
   - **ID**: 2349713
   - **Awardee**: University of Texas at Arlington
   - **Summary**: This project investigates whether animals have languages that are syntactically and semantically compositional like humans, using advanced machine learning.
   - **Program Officer**: Sharmistha Bagchi-Sen
   - **NSF Programs**: Research Experiences for Undergraduates (REU)

2. **Title**: STTR Phase I: Using Audio Analytics and Sensing to Enhance Broiler Chicken Welfare and Performance by Continuously Monitoring Bird Vocalizations
   - **ID**: 2335590
   - **Awardee**: AUDIOT, INC.
   - **Summary**: This project develops a smart monitoring system for chicken farms using audio analytics to enhance bird welfare by monitoring bird conditions.
   - **Program Officer**: Samir M. Iqbal
   - **NSF Programs**: Small Business Technology Transfer (STTR) Phase I

3. **Title**: Postdoctoral Fellowship: Do hotter climates promote the evolution of complex mating displays? Integrating physiology, behavior, & the macroevolution of communication systems
   - **ID**: 2410506
   - **Awardee**: University of California Berkeley
   - **Summary**: This project investigates how temperature shapes communication in species, focusing on the behavioral interactions and evolutionary changes owing to environmental shifts.
   - **Program Officer**: Andrea Holgado

4. **Title**: Postdoctoral Fellowship: PRFB: Characterizing the history of Batrachochytrium dendrobatidis, a globally-significant amphibian pathogen, with historic and sedimentary DNA
   - **ID**: 2410553
   - **Awardee**: University of California, Santa Cruz
   - **Summary**: This involves studying genetic interactions and communication patterns influenced by environmental factors.
   - **Program Officer**: Andrea Holgado

5. **Title**: MCA: Neural control of elaborate gestural displays
   - **ID**: 2423144
   - **Awardee**: Brown University
   - **Summary**: The project focuses on understanding the neural mechanisms controlling animal gestural movements related to communication.
   - **Program Officer**: Jodie Jawor

For your specific interest in processing animal sounds and language, you may want to reach out to the following program officers to discuss your proposal:

- **Sharmistha Bagchi-Sen** (Research Experiences for Undergraduates and related programs): shabagch@nsf.gov
- **Samir M. Iqbal** (STTR Phase I related programs): smiqbal@nsf.gov
- **Andrea Holgado** (Postdoctoral Fellowship, Biology, and related programs): aholgado@nsf.gov
- **Jodie Jawor** (Animal Behavior programs): jjawor@nsf.gov

For general inquiries or more precise guidance on your particular research interests, also consider contacting the Directorate for Biological Sciences at NSF, as they oversee many of the relevant programs. 

Taking these steps can help you refine your proposal and align it with NSF's current research priorities and initiatives. Good luck with your submission!

## Persona 2 – Ms. Sandy the General Public
- Sample prompt or information Ms. Sandy from the General Public may want to know
- “I want to know about the awards made to <abc> organization/university by area of science and award size. In addition, tell me if the Principal Investigators or Co-Principal Investigators on any of the awards appear on awards to other organizations, break that down by area of science as well.”

In [6]:
json_file_path = "awards_arlington.json"

# Load the API key 
with open(os.path.expanduser("~/.openai")) as f:
    line = f.readline().strip()
    if line.startswith("api_key="):
        api_key = line.split("=")[1]
if not api_key:
    raise ValueError("API key not found. Please set the OPENAI_API_KEY environment variable.")

In [7]:
# Create OpenAI client
client = openai.OpenAI(api_key=api_key)

# Load the text from the JSON file
appended_text = load_text_from_json(json_file_path)
org = "University of Texas at Arlington"
prompt = f"I want to know about the awards made to {org} by area of science and award size. In addition, tell me if the Principal Investigators or Co-Principal Investigators on any of the awards appear on awards to other organizations, break that down by area of science as well."

messages = [
    {
        "role": "system",
        "content": (
            "You are a helpful assistant who always follows requests exactly. "
            "You review the request and reflect upon your response before returning it to the user. "
            "The source data is from the NSF Award Search website. "
            "If the award data contains references, you should research those references and include the relevant information in your response."
            "If you use any references, please include them as citations in the response."
        )
    },
    {
        "role": "user",
        "content": f"{prompt}\nJSON data:{appended_text}"
    }
]

# Call GPT-4 and get the first response
response = call_gpt_4(client, messages)

print("AI Agent Response:")
print(response)

AI Agent Response:
The information is provided as requested, categorized by area of science, award size, and involvement of Principal Investigators or Co-Principal Investigators on awards to other organizations. 

### 1. Awards by Area of Science:

#### Electrical Engineering and Computer Science (EECS)
1. **Award Number 2154109**
   - **Title:** Fast 2D Beam Steering Device Integrated Directly on High Power VCSEL arrays
   - **Amount:** $400,000.00

2. **Award Number 2438700**
   - **Title:** EAGER: HCC: Exploring Human Factors: Is a Teleoperated Robotics Framework Feasible for Persons who are Visually Impaired?
   - **Amount:** $140,758.00

3. **Award Number 2425371**
   - **Title:** Collaborative Research: CIF: Small: Not All Eggs in One Basket: Authority Distribution for Resilience Against Compromised Nodes in Communication Networks
   - **Amount:** $240,058.00

4. **Award Number 2415174**
   - **Title:** Travel: Student Travel Support for the International Conference on Data Engin

AI Agent Response:
The information is provided as requested, categorized by area of science, award size, and involvement of Principal Investigators or Co-Principal Investigators on awards to other organizations. 

### 1. Awards by Area of Science:

#### Electrical Engineering and Computer Science (EECS)
1. **Award Number 2154109**
   - **Title:** Fast 2D Beam Steering Device Integrated Directly on High Power VCSEL arrays
   - **Amount:** $400,000.00

2. **Award Number 2438700**
   - **Title:** EAGER: HCC: Exploring Human Factors: Is a Teleoperated Robotics Framework Feasible for Persons who are Visually Impaired?
   - **Amount:** $140,758.00

3. **Award Number 2425371**
   - **Title:** Collaborative Research: CIF: Small: Not All Eggs in One Basket: Authority Distribution for Resilience Against Compromised Nodes in Communication Networks
   - **Amount:** $240,058.00

4. **Award Number 2415174**
   - **Title:** Travel: Student Travel Support for the International Conference on Data Engineering (ICDE) 2024
   - **Amount:** $25,000.00

5. **Award Number 2046632**
   - **Title:** CAREER: System Design of Crowd Logistics via Participatory Agent-based Modeling
   - **Amount:** $532,585.00

#### Mechanical Engineering
1. **Award Number 2422573**
   - **Title:** EAGER: Liutex-based Sub-Grid Model for Large Eddy Simulation of Turbulent Flow
   - **Amount:** $299,632.00

2. **Award Number 2349873**
   - **Title:** Collaborative Research: GEM--Multi-scale Magnetosphere-Ionosphere-Thermosphere Coupling Dynamics Driven by Bursty Bulk Flows
   - **Amount:** $148,179.00

#### Environmental Science
1. **Award Number 2004111**
   - **Title:** Single Molecule Fluorescent Indicators for Gas-Phase Sensing of Metals
   - **Amount:** $420,000.00

#### Chemistry and Chemical Engineering
1. **Award Number 2054353**
   - **Title:** Collaborative Research: Dynamical Tides in Close Stellar Binaries and Exoplanetary Systems
   - **Amount:** $481,538.00

2. **Award Number 2202598**
   - **Title:** Enhancing Active Learning in Additive Manufacturing Using a Bilingual, Assisted Virtual-Reality Platform
   - **Amount:** $837,138.00

3. **Award Number 2204536**
   - **Title:** SHF: Small: INCA: Incremental Analysis of Software Specification for Evolving Systems
   - **Amount:** $489,946.00

4. **Award Number 2401373**
   - **Title:** CAREER: Towards Privacy-Preserving Wireless Communication: Fundamental Limits and Coding Schemes
   - **Amount:** $344,375.00

### 2. Principal Investigators or Co-Principal Investigators Involvement in Other Organization's Awards by Area of Science:
#### Engineering
1. **Dereje Agonafer**:
   - **Awards:** '2209751', '2056489', '2324716'
   - **Organizations:** University of Texas at Arlington, Binghamton University

2. **Katariina Wartio**:
   - **Awards:** '2350744', '2350825'
   - **Organizations:** University of Oulu, University of Texas at Arlington

3. **Junha Jeon**:
   - **Awards:** '2102712', '1906213'
   - **Organizations:** University of Texas at Arlington, SUNY College of Environmental Science and Forestry

This dataset provides funding for varied scientific areas and demonstrates the cross-collaboration of researchers. The details above can help understand the funding spectrum provided to specific fields at the University of Texas at Arlington (UTA) and the potential cross-institutional collaboration contributing to these research endeavors.